### ISSUE TITLE
Garden not detecting finished buildkit job

### ISSUE THREAD

**Issue by to266** [garden-io/garden#3324]
### Current Behavior

Garden does not detect that a buildkit job is finished, and basically hangs. If the local CLI command is stopped and re-run, then it's correctly registered as already built, and the steps continue.

### Expected behavior

Correctly detect a finished build

### Reproducible example

(Note that I haven't yet tried to replicate this)
The dockerfile in particular is of the following format
```dockerfile
# The below get passed from other garden-build modules as buildargs
ARG BASE_IMAGE_VERSION
ARG BASE_IMAGE_NAME
FROM ${BASE_IMAGE_NAME}:${BASE_IMAGE_VERSION} as builder
WORKDIR /app
COPY . .
RUN ./command_that_builds.sh && \
  mkdir compiled-binaries && \
  ./copy_binaries.sh compiled-binaries && \
  ./clean_compilation.sh

# The below is only used as a binary container. I'd use scratch, but kaniko did not like that
FROM alpine
COPY --from=builder /app/compiled-binaries /
COPY --from=builder /app/resources /
```

### Your environment

* **OS:** Linux
* **How I'm running Kubernetes:** AKS

`garden version` 0.12.46

**Comment by to266** [comment]
(and the only way I was able to find out when the build is actually finished, is by running `top` inside the buildkit container and monitoring that)

**Comment by stefreak** [comment]
@to266 thanks for this report, this is very interesting. Sorry for this!

Can you run `garden` with the parameter `-l 5` for silly-log and provide us with the buildkit-related log output? Is this a regression in `0.12.46` or does it also happen with `0.12.45`?

**Comment by to266** [comment]
It used to happen with `0.12.45` as well, but before that the bigger blocker was the lack of ability to add tolerations.

In terms of silly logs, here's all there really is:
```
ℹ rust-release              → Syncing files to cluster...
   Forwarding local port 33547 to Pod/garden-buildkit-9d65d747b-vqfrp port 8730
   Running 'kubectl --context=lumiquote-dev --namespace=dev-tomas port-forward Pod/garden-buildkit-9d65d747b-vqfrp 33547:8730'
   [Pod/garden-buildkit-9d65d747b-vqfrp port forwarder] Forwarding from 127.0.0.1:33547 -> 8730
Forwarding from [::1]:33547 -> 8730

   Syncing from /home/tomas/lumi/epibator/.garden/build/./rust-release/ to rsync://localhost:33547/volume/82048baf-1ad6-409c-93de-e63965407bcb/
ℹ rust-release              → File sync to cluster complete
ℹ rust-release              → Building image rust-release:v-eb2b2d2764...
   Execing command in dev-tomas/Pod/garden-buildkit-9d65d747b-vqfrp/buildkitd: buildctl build --frontend=dockerfile.v0 --local context=/garden-build/82048baf-1ad6-409c-93de-e63965407bcb/rust-release --local dockerfile=/garden-build/82048baf-1ad6-409c-93de-e63965407bcb/rust-release --opt filename=docker/release.Dockerfile --output type=image,"name=lumiquote.azurecr.io/images/rust-release:v-eb2b2d2764",push=true --import-cache type=registry,ref=lumiquote.azurecr.io/images/rust-release:_buildcache-garden-buildkit --export-cache type=registry,ref=lumiquote.azurecr.io/images/rust-release:_buildcache-garden-buildkit,mode=max --import-cache type=registry,ref=lumiquote.azurecr.io/images/rust-release:_buildcache-master --opt build-arg:GARDEN_MODULE_VERSION=v-eb2b2d2764 --opt build-arg:BASE_IMAGE_VERSION=v-a15f756ccd --opt build-arg:BASE_IMAGE_NAME=lumiquote.azurecr.io/images/rust-workspace --opt target=output
  → #1 [internal] load build definition from docker/release.Dockerfile
  → #1 transferring dockerfile: 749B done
  → #1 DONE 0.0s
  →
  → #2 [internal] load .dockerignore
  → #2 transferring context: 63B done
  → #2 DONE 0.0s
  →
  → #4 [internal] load metadata for lumiquote.azurecr.io/images/rust-workspace:v-a15f756ccd
  → #4 ...
  →
  → #5 [auth] images/rust-workspace:pull token for lumiquote.azurecr.io
  → #5 DONE 0.0s
  →
  → #3 [internal] load metadata for docker.io/library/alpine:latest
  → #3 ...
  →
  → #4 [internal] load metadata for lumiquote.azurecr.io/images/rust-workspace:v-a15f756ccd
  → #4 DONE 0.7s
  →
  → #3 [internal] load metadata for docker.io/library/alpine:latest
  → #3 DONE 1.2s
  →
  → #11 [internal] load build context
  → #11 DONE 0.0s
  →
  → #9 [cacher 1/4] FROM lumiquote.azurecr.io/images/rust-workspace:v-a15f756ccd@sha256:a2181eb458f6ea9a51d3272ce0090f221cb2a92cf47e598043a9ec6fb083e0b0
  → #9 resolve lumiquote.azurecr.io/images/rust-workspace:v-a15f756ccd@sha256:a2181eb458f6ea9a51d3272ce0090f221cb2a92cf47e598043a9ec6fb083e0b0 done
  → #9 DONE 0.0s
  →
  → #8 [output 1/3] FROM docker.io/library/alpine@sha256:bc41182d7ef5ffc53a40b044e725193bc10142a1243f395ee852a8d9730fc2ad
  → #8 resolve docker.io/library/alpine@sha256:bc41182d7ef5ffc53a40b044e725193bc10142a1243f395ee852a8d9730fc2ad done
  → #8 DONE 0.0s
  →
  → #7 importing cache manifest from lumiquote.azurecr.io/images/rust-release:_buildcache-master
  → #7 ERROR: lumiquote.azurecr.io/images/rust-release:_buildcache-master: not found
  →
  → #16 [auth] images/rust-release:pull token for lumiquote.azurecr.io
  → #16 DONE 0.0s
  →
  → #17 [auth] images/rust-release:pull token for lumiquote.azurecr.io
  → #17 DONE 0.0s
  →
  → #6 importing cache manifest from lumiquote.azurecr.io/images/rust-release:_buildcache-garden-buildkit
  → #6 DONE 0.8s
  →
  → #11 [internal] load build context
  → #11 transferring context: 562.88kB 0.1s done
  → #11 DONE 0.1s
  →
  → #10 [cacher 2/4] WORKDIR /app
  → #10 CACHED
  →
  → #12 [cacher 3/4] COPY --chown=rust:rust . .
  → #12 DONE 0.1s
  →
  → #13 [cacher 4/4] RUN cargo build --release &&   mkdir compiled-binaries &&   find ./target/release -name "[^lib]*" -executable -type f -exec cp {} compiled-binaries/ ; &&   cargo clean
  → #13 0.635    Compiling proc_macros v0.1.0 (/app/proc_macros)
  → #13 0.641    Compiling auth-core v0.1.0 (/app/auth-core)
  → #13 0.641    Compiling scripting_language v0.1.0 (/app/scripting-language)
  → #13 0.644    Compiling api-docs v0.1.0 (/app/api-docs)
  → #13 0.646    Compiling epibator-utils v0.0.0 (/app/epibator-utils)
  → #13 0.649    Compiling ids v0.0.0 (/app/ids)
  → #13 0.652    Compiling tokener v0.1.0 (/app/tokener)
  → #13 0.861    Compiling primitives v0.0.0 (/app/primitives)
  → #13 0.986    Compiling template-renderer v0.1.0 (/app/template-renderer)
  → #13 1.852    Compiling segment_client v0.1.0 (/app/segment_client)
  → #13 2.722    Compiling ext-api-state v0.1.0 (/app/ext-api-state)
  → #13 3.581    Compiling tracing-utils v0.1.0 (/app/tracing-utils)
  → #13 4.557    Compiling persistent-cache v0.1.0 (/app/persistent-cache)
  → #13 5.121    Compiling auth-service v0.1.0 (/app/auth-service)
  → #13 5.128    Compiling domain-core v0.0.0 (/app/domain-core)
  → #13 21.09    Compiling tenant v0.0.0 (/app/tenant)
  → #13 21.19    Compiling notification-api v0.1.0 (/app/notification-api)
  → #13 22.01    Compiling part-parsing-utils v0.1.0 (/app/part-parsing-utils)
  → #13 22.82    Compiling bom-importer-client v0.1.0 (/app/bom-importer-client)
  → #13 23.92    Compiling formula-evaluator v0.1.0 (/app/formula-evaluator)
  → #13 25.71    Compiling domain-events v0.0.0 (/app/domain-events)
  → #13 27.63    Compiling webhook-receiver v0.1.0 (/app/webhook-receiver)
  → #13 31.33    Compiling magicbell-client v0.1.0 (/app/magicbell-client)
  → #13 31.37    Compiling octopart-part-search v0.1.0 (/app/octopart-part-search)
  → #13 34.37    Compiling notifications v0.1.0 (/app/notifications)
  → #13 35.88    Compiling epibator v0.0.0 (/app/epibator)
```

aka it stops at the compile step (technically does not even show the full output)

**Comment by to266** [comment]
If I split up the most expensive command

from 
```Dockerfile
RUN ./command_that_builds.sh && \
  mkdir compiled-binaries && \
  ./copy_binaries.sh compiled-binaries && \
  ./clean_compilation.sh
```
to
```Dockerfile
RUN ./command_that_builds.sh
RUN mkdir compiled-binaries && \
  ./copy_binaries.sh compiled-binaries && \
  ./clean_compilation.sh
```
it actually works correctly - but now in the build layer there's literally GBs of data that we don't need (yet that get cached)

**Comment by to266** [comment]
Huh, so what seems to have helped, is forcing `cargo` to output things in the same way that it would for an interactive TTY session via 
```Dockerfile
ENV CARGO_TERM_PROGRESS_WHEN=always
ENV CARGO_TERM_PROGRESS_WIDTH=120
```
[cargo docs](https://doc.rust-lang.org/cargo/reference/config.html).

Then the stdout output is somewhat mangled (as it assumes it can update the progressbar in-place), but at least it works :D

**Comment by stefreak** [comment]
@to266 no, absolutely not, we will dig into this and fix it. I think I also ran into this issue today in a simple dockerfile that just runs `apt-get update; apt-get install -y ...`

Thank you so much for sharing this with us

**Comment by stefreak** [comment]
@to266 Unfortunately I failed to reproduce that one, my initial suspicion that this happened to me too seems to have been wrong. 

I'm quite curious what is happeneing here, you are telling me that a workaround is to tell cargo that it logs to a tty. So I'm suspecting a bug in a component that handles log output.

Let's try to continue like this:

1. Can you tell me wether it still hangs if you redirect stdout / stderr to a file?
```
RUN (./command_that_builds.sh && \
  mkdir compiled-binaries && \
  ./copy_binaries.sh compiled-binaries && \
  ./clean_compilation.sh) 2>&1 > /output.txt
```
2. If it does not hang with output redirected to a file, can you share this file with me? I will then test if this can be reproduced with `cat /output.txt`

Thanks again for your time @to266

**Comment by agates4** [comment]
Actually, I have run into this too, and rerunning the original command fixes the hanging.
This happens on our end (Pachama) with large frontend builds. I haven't seen this problem with the slimmer backends

**Comment by stefreak** [comment]
We could reproduce this one with @agates4 using a `sleep 200` buildstep. 

It doesn't happen using kubectl exec, only with garden's javascript implementation equivalent to kubectl exec

We couldn't reproduce it in GKE.

My gut feeling is, that this is something on the HTTP or TCP layers, and happens with certain kinds of network setups. 

I suggest we try adding a ping to keep the websocket connection alive, see also https://websockets.readthedocs.io/en/stable/topics/timeouts.html#long-lived-connections

**Comment by agates4** [comment]
After manually testing with a built version of Garden from the linked pull request, the issue on our end is resolved!

**Comment by stefreak** [comment]
@to266 if you ever want to try again, this has been fixed since release [0.12.48](https://github.com/garden-io/garden/releases/tag/0.12.48) – again, tremendous thanks for reporting this in the first place.
